text/0000-ref-wrapping.md

```diff
@@ -0,0 +1,219 @@
+
+For the actual explanation: ref-wrapping allows `#[repr(transparent)]` literals to opportunistically inherit the lifetimes of the values in their primary field.
+
+When a `#[repr(transparent)]` literal is constructed, the type checker will work as usual, ensuring that the actual value of the primary field matches what's expected. However, the borrow checker will instead look at the literal as if it were the value of the primary field directly, rather than being moved into the literal. And since the type is transparent, the actual construction of the literal will be removed when computing the result, as it is today.
+
+When borrowing the literal, the same rules will apply for borrowing the field itself, e.g. you can't perform `&mut *value` if `value` is an immutable reference, etc. However, if `&*value` is allowed, then `&Transparent { primary_field: *value, .. }` is allowed, among similar constructs.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This is possible today with transmutes.
+* This may complicate the compiler for something that may be a relatively niche use case.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternative is to do nothing, and to force people to use unsafe code to achieve this. I personally disagree that this is acceptable, but it is a genuine alternative.
+
+The other alternative is to allow ref-wrapping via `as`-casting. For example, the `Path` example under this model could be done like the below:
+
+```rust
+fn from_bytes(s: &[u8]) -> &Path {
+    value as &Path
+}
+```
+
+This has a few problems:
+
+1. The language, as it stands, is moving away from `as` casts in favour of generic traits like `From`, and it's unclear how this could be replaced under this model. Deriving `From` seems prone to a bunch of errors, and potentially an alternative syntax would have to be introduced.
+2. This removes the semantics of constructing the type, including the values added to zero-sized fields. In effect, this is "revealing the hand" of the abstraction being a farce by not needing to construct the type. It could be argued that this is not a downside, but it feels worth including.
+3. It's unclear whether this should be allowed for values in addition to references, e.g. could `struct Seconds(u32)` allow casting `0 as Seconds`?
+4. It's less clear how this translates with privacy rules. Presumably, the privacy rules would be the same as this RFC's proposal, where it would only be allowed if the constructor is accessible, but that's still less clear than using the constructor directly.
+
+# Prior art
+[prior-art]: #prior-art
+
+As far as this RFC is concerned, there really isn't much prior art for this. Discussions have been had about using `as`-casting for this purpose, although as mentioned in the alternatives section, that has downsides.
```

Comment:
Prior Art: [the `ref_cast` crate](https://docs.rs/ref-cast/)

Reply:
if we're listing crates we could mention [bytemuck::TransparentWrapper](https://docs.rs/bytemuck/latest/bytemuck/trait.TransparentWrapper.html)

Reply:
Also [`aliri_braid`](https://docs.rs/aliri_braid) which is designed specifically for strongly-typed string wrappers.

Reply:
Also [`trapper`](https://crates.io/crates/trapper)